### PR TITLE
refactor: replace block_time based queies in address_tx_amount table …

### PR DESCRIPTION
**Pull request by Sotatek-DucPhung**
…with slot

### CODE REVIEW

File: src/main/resources/db/migration/ledgersyncagg/V1_4_6__drop_unused_indexes.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
+DROP INDEX IF EXISTS address_tx_amount_unit_block_time2_idx;
+DROP INDEX IF EXISTS address_tx_amount_unit_quantity_block_time_idx;
+DROP INDEX IF EXISTS address_tx_amount_unit_tx_hash_block_time_idx;
+DROP INDEX IF EXISTS address_tx_amount_address_tx_hash_block_time_idx;
+DROP INDEX IF EXISTS address_tx_amount_stake_address_tx_hash_block_time_idx;
+DROP INDEX IF EXISTS address_tx_amount_stake_address_unit_block_time_idx;
+DROP INDEX IF EXISTS address_tx_amount_slot_idx;
+
+CREATE INDEX IF NOT EXISTS address_tx_amount_block_time_idx
+    ON address_tx_amount (block_time);
```
Comment (nemo83) [inline src/main/resources/db/migration/ledgersyncagg/V1_4_6__drop_unused_indexes.sql:10]: do we still need index on `block_time`?

File: src/main/resources/db/migration/ledgersyncagg/V1_4_1__add-indexes.sql [hunk 2]
```diff
@@ -1,21 +1,9 @@
...
-CREATE INDEX IF NOT EXISTS address_tx_amount_address_tx_hash_block_time_idx
-    ON address_tx_amount (address, tx_hash, block_time);
-CREATE INDEX IF NOT EXISTS address_tx_amount_stake_address_tx_hash_block_time_idx
-    ON address_tx_amount (stake_address, tx_hash, block_time);
-CREATE INDEX IF NOT EXISTS address_tx_amount_stake_address_unit_block_time_idx
-    ON address_tx_amount (stake_address, unit, block_time);
-CREATE INDEX IF NOT EXISTS address_tx_amount_slot_idx
-    ON address_tx_amount (slot);
+CREATE INDEX IF NOT EXISTS address_tx_amount_block_time_idx
+    ON address_tx_amount (block_time);
```
Comment (nemo83) [inline src/main/resources/db/migration/ledgersyncagg/V1_4_1__add-indexes.sql:6]: do we need this?

Comment (nemo83) [reply src/main/resources/db/migration/ledgersyncagg/V1_4_1__add-indexes.sql:6]: @Kammerlo @Sotatek-DucPhung do you guys we could create an index here on

(unit, slot, quantity) to get top 20 ada/token holders for address (And do the same for stake_address_balance) ?

**Review by nemo83 — COMMENTED** [review]
left couple of comments, haven't reviewed fully.

File: src/main/resources/db/migration/ledgersyncagg/V1_4_1__add-indexes.sql (hunk 2 above)
Comment (Kammerlo) [reply src/main/resources/db/migration/ledgersyncagg/V1_4_1__add-indexes.sql:6]: We could do some tests to see if it increases the speed to something acceptable. Currently when calculating the top ada holders on these tables it will take minutes on mainnet. I'd give it a try.

File: src/main/java/org/cardanofoundation/job/repository/ledgersyncagg/AddressTxAmountRepository.java [hunk 3]
```diff
@@ -24,11 +24,11 @@ public interface AddressTxAmountRepository
   @Query(
       value =
           """
-      select (case when ata.stake_address is null then ata.address else ata.stake_address end) as account, count(distinct(ata.tx_hash)) as txCount
-      from address_tx_amount ata
-      where ata.epoch = :epochNo
-      and ata.slot != -1
-      group by account
+          select (case when ata.stake_address is null then ata.address else ata.stake_address end) as account, count(distinct(ata.tx_hash)) as txCount
```
Comment (Kammerlo) [inline src/main/java/org/cardanofoundation/job/repository/ledgersyncagg/AddressTxAmountRepository.java:27]: For code quality I prefer writing SQL Statements in upper case so that it's easier to read. 
And we are doing it in every other Query so just for continuity it is good to have the same style as well.

File: src/main/resources/db/migration/ledgersyncagg/V1_4_2__add-agg-materialized-view.sql [hunk 4]
```diff
@@ -26,15 +26,22 @@ GROUP BY ata.address, ata.stake_address, day
 ORDER BY day;
 ------------------------------------------------------------------------------------------------------------------------
 DROP TABLE IF EXISTS stake_tx_balance;
-CREATE MATERIALIZED VIEW IF NOT EXISTS stake_tx_balance AS
```
Comment (Kammerlo) [inline src/main/resources/db/migration/ledgersyncagg/V1_4_2__add-agg-materialized-view.sql:29]: Why did you remove the IF NOT EXISTS?

File: src/main/java/org/cardanofoundation/job/config/AppConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import org.cardanofoundation.conversions.ClasspathConversionsFactory;
+import org.cardanofoundation.conversions.domain.NetworkType;
+
+@Configuration
+public class AppConfiguration {
+
+  @Bean
+  public CardanoConverters cardanoConverters(@Value("${application.network}") String network) {
+    return switch (network) {
+      case "preprod" -> ClasspathConversionsFactory.createConverters(NetworkType.PREPROD);
```
[the remark this answers is not in the document]
Comment (Sotatek-DucPhung) [reply src/main/java/org/cardanofoundation/job/config/AppConfiguration.java]: Updated !

File: src/main/java/org/cardanofoundation/job/repository/ledgersyncagg/AddressTxAmountRepository.java (hunk 3 above)
Comment (Sotatek-DucPhung) [reply src/main/java/org/cardanofoundation/job/repository/ledgersyncagg/AddressTxAmountRepository.java]: Updated!

File: src/main/resources/db/migration/ledgersyncagg/V1_4_2__add-agg-materialized-view.sql (hunk 4 above)
Comment (Sotatek-DucPhung) [reply src/main/resources/db/migration/ledgersyncagg/V1_4_2__add-agg-materialized-view.sql]: I missed it when changing the MV query. Updated!
